metapool/sample_sheet.py

```diff
@@ -388,203 +365,566 @@ def merge(self, sheets):
+        # of the following columns. If the column existed previously it will
+        # be overwritten, otherwise it will be created here. Alternate versions
+        # of the column name have already been resolved at this point.
+
+        # Note that the amplicon notebook currently generates the same values
+        # for this column. If the functionality in the notebook changes, the
+        # output will continue to be redfined with the current values here.
+        well_description = table['Project Plate'].astype(str) + "." + table[
+            'Sample'].astype(str) + "." + table['Well'].astype(str)
+
+        if strict:
+            # legacy operation. All columns not defined in remapper will be
+            # filtered out.
+            out = table[self.remapper.keys()].copy()
+            out.rename(self.remapper, axis=1, inplace=True)
+        else:
+            out = table.copy(deep=True)
+
+            # if a column named 'index' is present in table, assume it is a
+            # numeric index and not a sequence of bases, which is required in
+            # the output. Assume the column that will become 'index' is
+            # defined in remapper.
+            if 'index' in set(out.columns):
+                out.drop(columns=['index'], inplace=True)
+
+            # if an alternate form of a column name defined in
+            # _KL_SAMPLE_SHEET_COLUMN_ALTS is found in table, assume it should
+            # be renamed to its proper form and be included in the output e.g.:
+            # 'sample_plate' -> 'Sample_Plate'.
+
+            # assume keys in _KL_SAMPLE_SHEET_COLUMN_ALTS do not overlap w/
+            # remapper (they currently do not). Define the full set of
+            # potential columns to rename in table.
+
+            # new syntax in 3.9 allows us to merge two dicts together w/OR.
+            remapper = KLSampleSheet.column_alts | self.remapper
+            out.rename(remapper, axis=1, inplace=True)
+
+            # out may contain additional columns that aren't allowed in the
+            # [Data] section of a sample-sheet e.g.: 'Extraction Kit Lot'.
+            # There may also be required columns that aren't defined in out.
+            subset = list(
+                set(self.data_columns) & set(
+                    out.columns))
+
+            out = out[subset]
+
+        # append the new 'Well_description' column, now that alternates have
+        # been removed and non-essential columns have been dropped.
+        out['Well_description'] = well_description
+
+        for column in self.data_columns:
+            if column not in out.columns:
+                warnings.warn('The column %s in the sample sheet is empty' %
+                              column)
+                out[column] = ''
+
+        return out
+
+    def _add_data_to_sheet(self, table, sequencer, lanes, assay, strict=True):
+        table = self._remap_table(table, strict)
+        if assay != _AMPLICON:
```

Comment:
If I understand this right, it looks like knowledge of a subclass sample sheet is being encoded in the base class? Should these types of operations be relegated the child they're relevant too?